dask/dataframe/io/orc/arrow.py

```diff
@@ -1,117 +1,634 @@
+from collections import defaultdict
+
+import numpy as np
+import pandas as pd
 import pyarrow as pa
 import pyarrow.orc as orc
+from fsspec.core import get_fs_token_paths
 
+from ....base import tokenize
+from ....delayed import Delayed
+from ....utils import apply
+from ..parquet.core import apply_filters
+from ..parquet.utils import _flatten_filters
 from ..utils import _get_pyarrow_dtypes, _meta_from_dtypes
+from .utils import ORCEngine, collect_files, collect_partitions
 
 
-class ArrowORCEngine:
+class ArrowORCEngine(ORCEngine):
     @classmethod
-    def read_metadata(
+    def get_dataset_info(
         cls,
-        fs,
-        paths,
-        columns,
-        index,
-        split_stripes,
-        aggregate_files,
-        **kwargs,
+        path,
+        columns=None,
+        index=None,
+        filters=None,
+        gather_statistics=True,
+        dataset_kwargs=None,
+        storage_options=None,
     ):
+        # ArrowORCEngine does not support optional
+        # dataset_kwargs yet. Raise an error to avoid
+        # silent failures.
+        if dataset_kwargs:
+            raise ValueError(
+                "Pyarrow ORC engine does not currently support "
+                "any 'dataset_kwargs' options."
+            )
 
-        # Convert root directory to file list.
-        # TODO: Handle hive-partitioned data
+        # Process file path(s)
+        fs, _, paths = get_fs_token_paths(
+            path, mode="rb", storage_options=storage_options or {}
+        )
+
+        # Generate a full list of files and
+        # directory partitions. We would like to use
+        # something like the pyarrow.dataset API to
+        # do this, but we will need to do it manually
+        # until ORC is supported upstream.
+        directory_partitions = []
+        directory_partition_keys = {}
         if len(paths) == 1 and not fs.isfile(paths[0]):
-            paths = fs.find(paths[0])
+            root_dir = paths[0]
+            paths = collect_files(root_dir, fs)
+            (
+                paths,
+                directory_partitions,
+                directory_partition_keys,
+            ) = collect_partitions(paths, root_dir, fs)
```

Comment:
@rjzamora would you be interested in letting pyarrow handle this part of the logic? (as it can do for pyarrow-dataset engine for parquet)

I know this is not yet possible with the released pyarrow, but I could collaborate on it to ensure we have ORC support in the dataset APIs in the next pyarrow release (scheduled half october, and we can already test it with the nightly releases before that). 
That's of course not a short-term solution, and I don't know how time sensitive this PR is, but it would avoid writing a lot of custom code in this PR that could become obsolete if pyarrow.datasets supported it.

Reply:
I just wanted to open an issue for it on the Arrow side, but I see you already did! -> https://issues.apache.org/jira/browse/ARROW-13572

Reply:
>would you be interested in letting pyarrow handle this part of the logic?

I would be **very** interested in this :)

>I know this is not yet possible with the released pyarrow, but I could collaborate on it to ensure we have ORC support in the dataset APIs in the next pyarrow release (scheduled half october, and we can already test it with the nightly releases before that).

As much as I don't  "want" to, I may need to *try* to merge code (like `collect_partitions`), even though I fully expect it to become obsolete very quickly.  I'd much rather rely on the pyarrow.dataset API here, but there is somewhat of an urgent need to get hive-partitioned ORC support out with the October cudf release (and using the October pyarrow release may be cutting things a bit close).  With that said, I do think it makes a lot of sense for us to start collaborating on the pyarrow.dataset-version of this code asap, and to make sure the place-holder `collect_partitions`-like logic is designed in such a way that pyarrow.dataset can be "dropped in" for pa.__version__ > 5.0.0 (or whatever the appropriate version cutoff may be).